crates/number/src/serialization.rs

```diff
@@ -1,69 +1,165 @@
+                    )?))
+                } else {
+                    Ok(U256(primitive_types::U256::from_dec_str(s).map_err(
+                        |err| E::custom(format!("failed to decode {s:?} as decimal u256: {err}")),
+                    )?))
+                }
+            }
+        }
+
+        deserializer.deserialize_str(U256Visitor)
     }
 }
 
-impl SerializeAs<U256> for HexOrDecimalU256 {
-    fn serialize_as<S>(source: &U256, serializer: S) -> Result<S::Ok, S::Error>
+impl SerializeAs<primitive_types::U256> for U256 {
+    fn serialize_as<S: Serializer>(
+        source: &primitive_types::U256,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error> {
+        serializer.serialize_str(&source.to_string())
+    }
+}
+
+impl Serialize for U256 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
     where
         S: Serializer,
     {
-        serialize(source, serializer)
+        U256::serialize_as(&self.0, serializer)
+    }
+}
+
+impl U256 {
+    pub fn zero() -> Self {
+        Self(primitive_types::U256::zero())
+    }
+
+    pub fn from_dec_str(value: &str) -> Result<Self, FromDecStrErr> {
+        Ok(Self(primitive_types::U256::from_dec_str(value)?))
     }
 }
 
-pub fn serialize<S>(value: &U256, serializer: S) -> Result<S::Ok, S::Error>
-where
-    S: Serializer,
-{
-    serializer.serialize_str(&value.to_string())
+impl PartialEq for U256 {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
 }
 
-pub fn deserialize<'de, D>(deserializer: D) -> Result<U256, D::Error>
-where
-    D: Deserializer<'de>,
-{
-    struct Visitor {}
-    impl<'de> de::Visitor<'de> for Visitor {
-        type Value = U256;
+impl PartialOrd for U256 {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
 
-        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
-            write!(
-                formatter,
-                "a u256 encoded either as 0x hex prefixed or decimal encoded string"
-            )
-        }
+impl Ord for U256 {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.0.cmp(&other.0)
+    }
+}
 
-        fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
-        where
-            E: de::Error,
-        {
-            if s.trim().starts_with("0x") {
-                U256::from_str_radix(s, 16).map_err(|err| {
-                    de::Error::custom(format!("failed to decode {s:?} as hex u256: {err}"))
-                })
-            } else {
-                U256::from_dec_str(s).map_err(|err| {
-                    de::Error::custom(format!("failed to decode {s:?} as decimal u256: {err}"))
-                })
-            }
-        }
+impl From<u128> for U256 {
+    fn from(value: u128) -> Self {
+        Self(primitive_types::U256::from(value))
+    }
+}
+
+impl From<u64> for U256 {
+    fn from(value: u64) -> Self {
+        Self(value.into())
+    }
+}
+
+impl From<i32> for U256 {
```

Comment:
Converting an signed type to an unsigned one without a way to notice that we transformed a negative number seems dangerous. I think if there is ever the need to convert from a singed type it's okay to make the developer jump through some hoops to proof to the compiler that this is safe.
Alternatively you could implement `TryFrom` instead.

Reply:
I totally agree with you. That implementation is the one coming from the `primitive_types`. The `From<i32>` was only used for testing purposes. I changed the tests to use `unsigned`. Thanks!